src/commands.c

```diff
@@ -827,6 +833,368 @@ int yh_com_generate_asymmetric(yubihsm_context *ctx, Argument *argv,
+
+    uint8_t signatureBytes[512]; // current max signature size (RSA4096)
+    size_t signature_len = sizeof(signatureBytes);;
+    EVP_PKEY *pubkey = X509_REQ_get0_pubkey(csr); // not to be freed
+    yh_rc yrc = 0;
+    if (EVP_PKEY_is_a(pubkey, "RSA")) {
+        yrc = yh_util_sign_pkcs1v1_5(session, key_id, true, hashed_data, hashed_data_len, signatureBytes, &signature_len);
+    } else if (EVP_PKEY_is_a(pubkey, "EC")) {
+        yrc = yh_util_sign_ecdsa(session, key_id, hashed_data, hashed_data_len, signatureBytes, &signature_len);
+    } else if (EVP_PKEY_is_a(pubkey, "ED25519")) { // Requires OpenSSL 1.1.1
+        yrc = yh_util_sign_eddsa(session, key_id, hashed_data, hashed_data_len, signatureBytes, &signature_len);
+    } else {
+        fprintf( stderr, "key algo must be RSA, EC, or ED");
+    }
+    if (yrc != YHR_SUCCESS) {
+        fprintf(stderr, "Failed to sign CSR: %s\n", yh_strerror(yrc));
+        return NULL;
+    }
+
+    // Prepare the signature
+    ASN1_BIT_STRING *signature = ASN1_BIT_STRING_new();
+    ASN1_BIT_STRING_set(signature, signatureBytes, signature_len);
+    // When OpenSSL writes an ASN1_BIT_STRING into DER, it encodes the number of bits left in the first octet
+    // CSR signatures are byte-alligned: so first fix this to indicate 0 bits left
+    signature->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07); // clear
+    signature->flags |= ASN1_STRING_FLAG_BITS_LEFT; // no unused bits
+    return signature;
+}
+
+// NOTE: Generate a Certificate Signing Request
+// argc = 3
+// arg 0: e:session
+// arg 1: w:key_id
+// arg 2: s:subject
+int yh_com_generate_csr(yubihsm_context *ctx, Argument *argv, cmd_format in_fmt,
+                      cmd_format fmt) {
+
+  UNUSED(in_fmt);
+
+  EVP_PKEY *public_key = NULL;
+  X509_ALGOR *alg = NULL;
+  X509_REQ *csr = NULL;
+  int ret = -1;
+
+  if ((csr = X509_REQ_new()) == NULL) {
+    fprintf(stderr, "Failed to generate CSR\n");
+    return -1;
+  }
+  if (!X509_REQ_set_version(csr, X509_REQ_VERSION_1)) {
+    fprintf(stderr, "Failed to generate CSR\n");
+    ret = -1;
+    goto csr_cleanup;
+  }
+  public_key = get_public_key(argv[0].e, argv[1].w);
+  if(!public_key) {
+    fprintf(stderr, "Failed to retrieve public key\n");
+    ret = -1;
+    goto csr_cleanup;
+  }
+  if (!X509_REQ_set_pubkey(csr, public_key)) {
+    fprintf(stderr, "Failed to generate CSR\n");
+    ret = -1;
+    goto csr_cleanup;
+  }
+  X509_NAME *name = parse_name(argv[2].s);
+  if(!name) {
+    fprintf(stderr, "Failed to parse subject name\n");
+    ret = -1;
+    goto csr_cleanup;
+  }
+  if(!X509_REQ_set_subject_name(csr, name)) {
+    fprintf(stderr, "Failed to generate CSR\n");
+    ret = -1;
+    goto csr_cleanup;
+  }
+
+  ASN1_BIT_STRING *signature = calculate_self_signature(argv[0].e, argv[1].w, csr);
+  if(!signature) {
+    fprintf(stderr, "Failed to generate CSR signature\n");
+    ret = -1;
+    goto csr_cleanup;
+  }
+  X509_REQ_set0_signature(csr, signature);
```

Comment:
The X509_REQ_set0_signature() and X509_REQ_set1_signature_algo() were added in OpenSSL 1.1.1e - causing build issues on debain:11 and ubuntu 20.04